server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java

```diff
@@ -119,6 +120,8 @@ public final class ThreadContext implements Writeable {
     private final long maxWarningHeaderSize;
     private final List<ThreadContextStatePropagator> propagators;
 
+    private static final Permission ACCESS_SYSTEM_THREAD_CONTEXT_PERMISSION = new RuntimePermission("markAsSystemContext");
```

Comment:
Does that mean the check should not be in the first few lines of `ThreadContext.markAsSystemContext`? I can write a new method within SecureSM, but I'm not sure how to call it.

Reply:
I think the check is fine here. On a closer look, so we have this `ThreadPermission` class and it would be good to introduce  `ThreadContextPermission` along with it.

Reply:
Pushed a change to introduce `ThreadContextPermission`. Let me know what you think. I'm waiting for CI results to ensure this change doesn't break any tests. I need to step away for a few hours, but I'll make any further changes once back.